File: src/python/airlock_gateway/auth_client.py
import time
import json
import hashlib
import os
import base64
import asyncio
from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib.parse import urlencode, urlparse, parse_qs
from datetime import datetime, timedelta, timezone
from typing import Optional, Callable, Dict, Any, Tuple
from dataclasses import dataclass
import httpx
import threading

from .models import DeviceCodeInfo, TokenResponse, ConsentErrorInfo
from .exceptions import AirlockAuthException

class AirlockAuthOptions:
    """Configuration for AirlockAuthClient."""
    def __init__(self, keycloak_realm_url: str, oidc_client_id: str = "airlock-integrations"):
        self.keycloak_realm_url = keycloak_realm_url
        self.oidc_client_id = oidc_client_id

class OidcDiscoveryResult:
    def __init__(self, data: Dict[str, Any]):
        self.token_endpoint = data.get("token_endpoint")
        self.device_authorization_endpoint = data.get("device_authorization_endpoint")
        self.revocation_endpoint = data.get("revocation_endpoint")
        self.authorization_endpoint = data.get("authorization_endpoint")

class AirlockAuthClient:
    """
    Handles user authentication for enforcer apps.
    Supports two OAuth2 flows:
      - Device Authorization Grant (RFC 8628) — for headless/CLI apps (Agent, Desktop, VS Code Extension)
      - Authorization Code + PKCE (RFC 7636) — for browser-capable apps (Web, Mobile)
    """
    def __init__(self, options: AirlockAuthOptions, httpx_client: Optional[httpx.AsyncClient] = None):
        self.options = options
        self._http = httpx_client or httpx.AsyncClient(timeout=30.0)
        
        self._oidc_config: Optional[OidcDiscoveryResult] = None
        self._access_token: Optional[str] = None
        self._refresh_token: Optional[str] = None
        self._token_expires_at: float = 0.0

    async def _close(self):
        await self._http.aclose()

    @property
    def current_access_token(self) -> Optional[str]:
        return self._access_token

    @property
    def is_logged_in(self) -> bool:
        return self._access_token is not None

    @property
    def is_token_expired(self) -> bool:
        return datetime.now(timezone.utc).timestamp() >= self._token_expires_at

    async def discover(self) -> OidcDiscoveryResult:
        if self._oidc_config:
            return self._oidc_config

        realm_url = self.options.keycloak_realm_url.rstrip("/")
        well_known_url = f"{realm_url}/.well-known/openid-configuration"

        resp = await self._http.get(well_known_url)
        resp.raise_for_status()

        data = resp.json()
        self._oidc_config = OidcDiscoveryResult(data)

        if not self._oidc_config.token_endpoint:
            raise RuntimeError("OIDC discovery: token_endpoint is missing.")

        return self._oidc_config

    async def login(self, on_user_code: Callable[[DeviceCodeInfo], None]) -> TokenResponse:
        """
        Start the Device Authorization Grant flow.
        Returns device code info (user_code, verification_uri) for the user to complete in their browser.
        Then polls the token endpoint until the user authorizes.
        """
        oidc = await self.discover()

        # Step 1: Request device code
        device_resp = await self._http.post(
            oidc.device_authorization_endpoint,
            data={
                "client_id": self.options.oidc_client_id,
                "scope": "openid profile email"
            }
        )
        device_resp.raise_for_status()
        device_data = device_resp.json()
        
        device_code = DeviceCodeInfo(device_data)

        # Step 2: Notify caller to display the code
        on_user_code(device_code)

        # Step 3: Poll token endpoint
        interval = max(device_code.interval, 5)
        deadline = datetime.now(timezone.utc).timestamp() + device_code.expires_in

        while datetime.now(timezone.utc).timestamp() < deadline:
            time.sleep(interval)

            token_resp = await self._http.post(
                oidc.token_endpoint,
                data={
                    "grant_type": "urn:ietf:params:oauth:grant-type:device_code",
                    "client_id": self.options.oidc_client_id,
                    "device_code": device_code.device_code
                }
            )

            token_data = token_resp.json()

            if token_resp.is_success:
                token = TokenResponse(token_data)
                self._access_token = token.access_token
                self._refresh_token = token.refresh_token
                self._token_expires_at = datetime.now(timezone.utc).timestamp() + token.expires_in - 30
                return token

            error = token_data.get("error")
            if error == "authorization_pending":
                continue
            elif error == "slow_down":
                interval += 5
                continue
            elif error == "expired_token":
                raise AirlockAuthException("Device code expired. Please try logging in again.")
            elif error == "access_denied":
                raise AirlockAuthException("User denied the authorization request.")
            else:
                desc = token_data.get("error_description", "")
                raise AirlockAuthException(f"Token request failed: {error} — {desc}")

        raise AirlockAuthException("Device code expired before user completed authorization.")

    # ── Authorization Code + PKCE ───────────────────────────────

    async def login_with_auth_code(
        self,
        on_browser_url: Callable[[str], None],
        redirect_port: int = 0
    ) -> TokenResponse:
        """
        Start the Authorization Code + PKCE flow.
        Opens a local HTTP server to receive the callback, then exchanges the code for tokens.
        Best for Web and Mobile enforcer apps.
        """
        oidc = await self.discover()
        if not oidc.authorization_endpoint:
            raise RuntimeError("OIDC discovery: authorization_endpoint is missing.")

        # Step 1: Generate PKCE code_verifier + code_challenge
        code_verifier = self._generate_code_verifier()
        code_challenge = self._compute_code_challenge(code_verifier)
        state = os.urandom(16).hex()

        # Step 2: Find an available port if not specified
        if redirect_port == 0:
            import socket
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
                s.bind(('127.0.0.1', 0))
                redirect_port = s.getsockname()[1]

        redirect_uri = f"http://localhost:{redirect_port}/callback"

        # Step 3: Build authorization URL
        params = urlencode({
            "response_type": "code",
            "client_id": self.options.oidc_client_id,
            "redirect_uri": redirect_uri,
            "scope": "openid profile email",
            "state": state,
            "code_challenge": code_challenge,
            "code_challenge_method": "S256",
        })
        auth_url = f"{oidc.authorization_endpoint}?{params}"

        # Step 4: Start local HTTP server and wait for callback
        result: Dict[str, Any] = {}
        error_info: Dict[str, str] = {}

        class CallbackHandler(BaseHTTPRequestHandler):
            def do_GET(handler_self):
                parsed = urlparse(handler_self.path)
                if parsed.path != "/callback":
                    handler_self.send_response(404)
                    handler_self.end_headers()
                    return

                params = parse_qs(parsed.query)

                err = params.get("error", [None])[0]
                if err:
                    error_info["error"] = err
                    error_info["description"] = params.get("error_description", [""])[0]
                    handler_self.send_response(400)
                    handler_self.send_header("Content-Type", "text/html")
                    handler_self.end_headers()
                    handler_self.wfile.write(b"Authorization failed. You can close this tab.")
                    return

                returned_state = params.get("state", [None])[0]
                if returned_state != state:
                    error_info["error"] = "state_mismatch"
                    handler_self.send_response(400)
                    handler_self.send_header("Content-Type", "text/html")
                    handler_self.end_headers()
                    handler_self.wfile.write(b"Invalid state parameter.")
                    return

                code = params.get("code", [None])[0]
                if code:
                    result["code"] = code
                    handler_self.send_response(200)
                    handler_self.send_header("Content-Type", "text/html")
                    handler_self.end_headers()
                    handler_self.wfile.write(b"Authorization successful! You can close this tab.")

            def log_message(self, format, *args):
                pass  # Suppress logs

        server = HTTPServer(('127.0.0.1', redirect_port), CallbackHandler)
        server.timeout = 300  # 5 minute timeout

        on_browser_url(auth_url)

        # Handle one request in a thread to not block async
        server_thread = threading.Thread(target=server.handle_request)
        server_thread.start()
        server_thread.join(timeout=300)
        server.server_close()

        if error_info:
            raise AirlockAuthException(
                f"Authorization denied: {error_info.get('error')} — {error_info.get('description', '')}")

        if "code" not in result:
            raise AirlockAuthException("Authorization timed out or no code received.")

        # Step 5: Exchange code for tokens
        return await self.exchange_code(result["code"], redirect_uri, code_verifier)

    async def get_authorization_url(self, redirect_uri: str) -> Dict[str, str]:
        """
        Builds the authorization URL for the Auth Code + PKCE flow.
        Use this when you manage the browser redirect yourself.
        Returns dict with: authorization_url, state, code_verifier, redirect_uri.
        """
        oidc = await self.discover()
        if not oidc.authorization_endpoint:
            raise RuntimeError("OIDC discovery: authorization_endpoint is missing.")

        code_verifier = self._generate_code_verifier()
        code_challenge = self._compute_code_challenge(code_verifier)
        state = os.urandom(16).hex()

        params = urlencode({
            "response_type": "code",
            "client_id": self.options.oidc_client_id,
            "redirect_uri": redirect_uri,
            "scope": "openid profile email",
            "state": state,
            "code_challenge": code_challenge,
            "code_challenge_method": "S256",
        })

        return {
            "authorization_url": f"{oidc.authorization_endpoint}?{params}",
            "state": state,
            "code_verifier": code_verifier,
            "redirect_uri": redirect_uri,
        }

    async def exchange_code(self, code: str, redirect_uri: str, code_verifier: str) -> TokenResponse:
        """Exchange an authorization code for tokens (Auth Code + PKCE)."""
        oidc = await self.discover()

        resp = await self._http.post(
            oidc.token_endpoint,
            data={
                "grant_type": "authorization_code",
                "client_id": self.options.oidc_client_id,
                "code": code,
                "redirect_uri": redirect_uri,
                "code_verifier": code_verifier,
            }
        )

        data = resp.json()
        if not resp.is_success:
            error = data.get("error", "")
            desc = data.get("error_description", "")
            raise AirlockAuthException(f"Code exchange failed: {error} — {desc}")

        token = TokenResponse(data)
        self._access_token = token.access_token
        self._refresh_token = token.refresh_token
        self._token_expires_at = datetime.now(timezone.utc).timestamp() + token.expires_in - 30
        return token

    @staticmethod
    def _generate_code_verifier() -> str:
        return base64.urlsafe_b64encode(os.urandom(32)).rstrip(b'=').decode('ascii')

    @staticmethod
    def _compute_code_challenge(code_verifier: str) -> str:
        digest = hashlib.sha256(code_verifier.encode('ascii')).digest()
        return base64.urlsafe_b64encode(digest).rstrip(b'=').decode('ascii')

    async def refresh_token(self) -> TokenResponse:
        if not self._refresh_token:
            raise AirlockAuthException("No refresh token available. Please login first.")

        oidc = await self.discover()

        resp = await self._http.post(
            oidc.token_endpoint,
            data={
                "grant_type": "refresh_token",
                "client_id": self.options.oidc_client_id,
                "refresh_token": self._refresh_token
            }
        )

        data = resp.json()

        if not resp.is_success:
            self._access_token = None
            self._refresh_token = None
            error = data.get("error")
            desc = data.get("error_description", "")
            raise AirlockAuthException(f"Token refresh failed: {error} — {desc}")

        token = TokenResponse(data)
        self._access_token = token.access_token
        self._refresh_token = token.refresh_token
        self._token_expires_at = datetime.now(timezone.utc).timestamp() + token.expires_in - 30
        return token

    async def get_access_token(self) -> str:
        if not self._access_token:
            raise AirlockAuthException("Not logged in. Call login() first.")

        if self.is_token_expired and self._refresh_token:
            await self.refresh_token()

        if not self._access_token:
            raise AirlockAuthException("Token refresh failed and no valid token available.")

        return self._access_token

    async def logout(self):
        if self._refresh_token:
            try:
                oidc = await self.discover()
                if oidc.revocation_endpoint:
                    await self._http.post(
                        oidc.revocation_endpoint,
                        data={
                            "client_id": self.options.oidc_client_id,
                            "token": self._refresh_token,
                            "token_type_hint": "refresh_token"
                        }
                    )
            except Exception:
                pass # Best-effort revocation

        self._access_token = None
        self._refresh_token = None
        self._token_expires_at = 0.0

    @staticmethod
    def parse_consent_error(status_code: int, response_body: str) -> Optional[ConsentErrorInfo]:
        if status_code != 403:
            return None

        try:
            data = json.loads(response_body)
            error = data.get("error")
            
            if error not in ("app_consent_required", "app_consent_pending", "app_consent_denied"):
                return None

            return ConsentErrorInfo(data)
        except Exception:
            return None

    def restore_tokens(self, access_token: str, refresh_token: str, expires_at: float):
        self._access_token = access_token
        self._refresh_token = refresh_token
        self._token_expires_at = expires_at

    def get_token_state(self) -> Tuple[Optional[str], Optional[str], float]:
        return (self._access_token, self._refresh_token, self._token_expires_at)

File: src/typescript/src/auth-client.ts
/**
 * Options for configuring the AirlockAuthClient.
 */
export interface AirlockAuthOptions {
  /**
   * The Keycloak realm URL (e.g., https://auth.airlocks.io/realms/airlock)
   * Used for OIDC discovery and Device Authorization Grant.
   */
  keycloakRealmUrl: string;

  /**
   * The OIDC client ID for device authorization.
   * Default: "airlock-integrations" (the public client configured in Keycloak).
   */
  oidcClientId?: string;

  /**
   * Custom fetch implementation (optional, used for testing or overriding globals).
   */
  fetch?: typeof fetch;
}

export interface OidcDiscoveryResult {
  token_endpoint?: string;
  device_authorization_endpoint?: string;
  revocation_endpoint?: string;
  authorization_endpoint?: string;
}

export interface AuthCodeRequest {
  /** The full authorization URL to redirect the user to. */
  authorizationUrl: string;
  /** CSRF protection state parameter (validate on callback). */
  state: string;
  /** PKCE code_verifier — pass to exchangeCode(). */
  codeVerifier: string;
  /** The redirect URI used in the request. */
  redirectUri: string;
}

export interface DeviceCodeInfo {
  device_code: string;
  user_code: string;
  verification_uri: string;
  verification_uri_complete?: string;
  expires_in: number;
  interval: number;
}

export interface TokenResponse {
  access_token: string;
  refresh_token: string;
  token_type: string;
  expires_in: number;
  scope: string;
}

export interface ConsentErrorInfo {
  error: string;
  message?: string;
  consentUrl?: string;
  appName?: string;
  appId?: string;
}

/**
 * Handles user authentication for enforcer apps.
 * Supports two OAuth2 flows:
 * - Device Authorization Grant (RFC 8628) — for headless/CLI apps (Agent, Desktop, VS Code Extension)
 * - Authorization Code + PKCE (RFC 7636) — for browser-capable apps (Web, Mobile)
 */
export class AirlockAuthClient {
  private readonly options: AirlockAuthOptions;
  private readonly fetchFn: typeof fetch;

  private oidcConfig: OidcDiscoveryResult | null = null;
  private accessToken: string | null = null;
  private refreshToken: string | null = null;
  private tokenExpiresAt: number = 0; // Epoch timestamp (ms)

  constructor(options: AirlockAuthOptions) {
    this.options = { ...options, oidcClientId: options.oidcClientId || 'airlock-integrations' };
    this.fetchFn = options.fetch || fetch;
  }

  /** Gets the current access token (may be null if not logged in). */
  get currentAccessToken(): string | null {
    return this.accessToken;
  }

  /** True if the user is currently logged in with a valid token. */
  get isLoggedIn(): boolean {
    return this.accessToken !== null;
  }

  /** True if the access token has expired (refresh needed). */
  get isTokenExpired(): boolean {
    return Date.now() >= this.tokenExpiresAt;
  }

  // ── OIDC Discovery ──────────────────────────────────────────

  /**
   * Discovers OIDC endpoints from Keycloak's well-known configuration.
   * Called automatically by login() if not already done.
   */
  async discover(): Promise<OidcDiscoveryResult> {
    if (this.oidcConfig) return this.oidcConfig;

    const realmUrl = this.options.keycloakRealmUrl.replace(/\/$/, '');
    const wellKnownUrl = `${realmUrl}/.well-known/openid-configuration`;

    const resp = await this.fetchFn(wellKnownUrl);
    if (!resp.ok) {
      throw new Error(`Failed to fetch OIDC discovery document: ${resp.status} ${resp.statusText}`);
    }

    this.oidcConfig = (await resp.json()) as OidcDiscoveryResult;

    if (!this.oidcConfig.token_endpoint) {
      throw new Error('OIDC discovery: token_endpoint is missing.');
    }

    return this.oidcConfig;
  }

  // ── Device Authorization Grant ──────────────────────────────

  /**
   * Start the Device Authorization Grant flow.
   * Returns device code info (user_code, verification_uri) for the user to complete in their browser.
   * Then polls the token endpoint until the user authorizes.
   *
   * @param onUserCode Callback invoked when the user must open a URL and enter a code.
   *                   Display the verification_uri_complete or verification_uri + user_code to the user.
   * @param abortSignal AbortSignal to cancel polling.
   * @returns The token response with access_token and refresh_token.
   */
  async login(
    onUserCode: (info: DeviceCodeInfo) => void,
    abortSignal?: AbortSignal
  ): Promise<TokenResponse> {
    const oidc = await this.discover();

    // Step 1: Request device code
    const deviceBody = new URLSearchParams({
      client_id: this.options.oidcClientId!,
      scope: 'openid profile email',
    });

    const deviceResp = await this.fetchFn(oidc.device_authorization_endpoint!, {
      method: 'POST',
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      body: deviceBody.toString(),
      signal: abortSignal,
    });

    if (!deviceResp.ok) {
      throw new Error(`Device authorization request failed: ${deviceResp.status}`);
    }

    const deviceCode = (await deviceResp.json()) as DeviceCodeInfo;

    // Step 2: Notify caller to display the code
    onUserCode(deviceCode);

    // Step 3: Poll token endpoint
    let interval = Math.max(deviceCode.interval || 5, 5);
    const deadline = Date.now() + (deviceCode.expires_in * 1000);

    while (Date.now() < deadline) {
      if (abortSignal?.aborted) {
        throw new Error('Login cancelled by user.');
      }

      await new Promise(resolve => setTimeout(resolve, interval * 1000));

      const tokenBody = new URLSearchParams({
        grant_type: 'urn:ietf:params:oauth:grant-type:device_code',
        client_id: this.options.oidcClientId!,
        device_code: deviceCode.device_code,
      });

      const tokenResp = await this.fetchFn(oidc.token_endpoint!, {
        method: 'POST',
        headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
        body: tokenBody.toString(),
        signal: abortSignal,
      });

      const tokenJson = await tokenResp.json();

      if (tokenResp.ok) {
        const token = tokenJson as TokenResponse;
        this.accessToken = token.access_token;
        this.refreshToken = token.refresh_token;
        this.tokenExpiresAt = Date.now() + ((token.expires_in - 30) * 1000);
        return token;
      }

      // Handle expected polling errors
      switch (tokenJson.error) {
        case 'authorization_pending':
          continue; // User hasn't authorized yet, keep polling
        case 'slow_down':
          interval += 5; // Back off
          continue;
        case 'expired_token':
          throw new Error('Device code expired. Please try logging in again.');
        case 'access_denied':
          throw new Error('User denied the authorization request.');
        default:
          throw new Error(`Token request failed: ${tokenJson.error} — ${tokenJson.error_description}`);
      }
    }

    throw new Error('Device code expired before user completed authorization.');
  }

  // ── Authorization Code + PKCE ──────────────────────────────

  /**
   * Start the Authorization Code + PKCE flow.
   * Opens a local HTTP server to receive the callback, then exchanges the code for tokens.
   * Best for Web and Mobile enforcer apps.
   *
   * @param onBrowserUrl Callback invoked with the authorization URL for the user to open.
   * @param redirectPort Local port for the callback listener (default: 0 = auto-select).
   * @param abortSignal AbortSignal to cancel the flow.
   */
  async loginWithAuthCode(
    onBrowserUrl: (url: string) => void,
    redirectPort: number = 0,
    abortSignal?: AbortSignal
  ): Promise<TokenResponse> {
    const oidc = await this.discover();
    if (!oidc.authorization_endpoint) {
      throw new Error('OIDC discovery: authorization_endpoint is missing.');
    }

    const { createServer } = await import('http');
    const { randomBytes, createHash } = await import('crypto');

    // Step 1: Generate PKCE code_verifier + code_challenge
    const codeVerifier = randomBytes(32).toString('base64url');
    const codeChallenge = createHash('sha256').update(codeVerifier).digest('base64url');
    const state = randomBytes(16).toString('hex');

    // Step 2: Start local HTTP server
    return new Promise<TokenResponse>((resolve, reject) => {
      const server = createServer(async (req, res) => {
        try {
          const url = new URL(req.url || '/', `http://localhost:${redirectPort}`);
          if (url.pathname !== '/callback') {
            res.writeHead(404);
            res.end();
            return;
          }

          const error = url.searchParams.get('error');
          if (error) {
            const desc = url.searchParams.get('error_description') || '';
            this.sendCallbackHtml(res, false, `Authorization failed: ${error}`);
            server.close();
            reject(new Error(`Authorization denied: ${error} — ${desc}`));
            return;
          }

          const returnedState = url.searchParams.get('state');
          if (returnedState !== state) {
            this.sendCallbackHtml(res, false, 'Invalid state parameter.');
            server.close();
            reject(new Error('CSRF state mismatch.'));
            return;
          }

          const code = url.searchParams.get('code');
          if (!code) {
            this.sendCallbackHtml(res, false, 'No authorization code received.');
            server.close();
            reject(new Error('No authorization code received in callback.'));
            return;
          }

          this.sendCallbackHtml(res, true, 'Authorization successful! You can close this tab.');

          const redirectUri = `http://localhost:${(server.address() as any).port}/callback`;
          const token = await this.exchangeCode(code, redirectUri, codeVerifier);
          server.close();
          resolve(token);
        } catch (e) {
          server.close();
          reject(e);
        }
      });

      server.listen(redirectPort, '127.0.0.1', () => {
        const port = (server.address() as any).port;
        const redirectUri = `http://localhost:${port}/callback`;

        const authUrl = `${oidc.authorization_endpoint}` +
          `?response_type=code` +
          `&client_id=${encodeURIComponent(this.options.oidcClientId!)}` +
          `&redirect_uri=${encodeURIComponent(redirectUri)}` +
          `&scope=${encodeURIComponent('openid profile email')}` +
          `&state=${state}` +
          `&code_challenge=${codeChallenge}` +
          `&code_challenge_method=S256`;

        onBrowserUrl(authUrl);
      });

      // Timeout after 5 minutes
      const timeout = setTimeout(() => {
        server.close();
        reject(new Error('Authorization timed out (5 minutes).'));
      }, 5 * 60 * 1000);

      server.on('close', () => clearTimeout(timeout));

      if (abortSignal) {
        abortSignal.addEventListener('abort', () => {
          server.close();
          reject(new Error('Login cancelled by user.'));
        });
      }
    });
  }

  /**
   * Builds the authorization URL for the Auth Code + PKCE flow.
   * Use this when you manage the browser redirect yourself (e.g. in a web app).
   * After the user authorizes, call exchangeCode() with the returned code.
   */
  async getAuthorizationUrl(redirectUri: string): Promise<AuthCodeRequest> {
    const oidc = await this.discover();
    if (!oidc.authorization_endpoint) {
      throw new Error('OIDC discovery: authorization_endpoint is missing.');
    }

    const { randomBytes, createHash } = await import('crypto');
    const codeVerifier = randomBytes(32).toString('base64url');
    const codeChallenge = createHash('sha256').update(codeVerifier).digest('base64url');
    const state = randomBytes(16).toString('hex');

    const authorizationUrl = `${oidc.authorization_endpoint}` +
      `?response_type=code` +
      `&client_id=${encodeURIComponent(this.options.oidcClientId!)}` +
      `&redirect_uri=${encodeURIComponent(redirectUri)}` +
      `&scope=${encodeURIComponent('openid profile email')}` +
      `&state=${state}` +
      `&code_challenge=${codeChallenge}` +
      `&code_challenge_method=S256`;

    return { authorizationUrl, state, codeVerifier, redirectUri };
  }

  /**
   * Exchange an authorization code for tokens (used with Auth Code + PKCE flow).
   */
  async exchangeCode(code: string, redirectUri: string, codeVerifier: string): Promise<TokenResponse> {
    const oidc = await this.discover();

    const body = new URLSearchParams({
      grant_type: 'authorization_code',
      client_id: this.options.oidcClientId!,
      code,
      redirect_uri: redirectUri,
      code_verifier: codeVerifier,
    });

    const resp = await this.fetchFn(oidc.token_endpoint!, {
      method: 'POST',
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      body: body.toString(),
    });

    const json = await resp.json();

    if (!resp.ok) {
      throw new Error(`Code exchange failed: ${json.error} — ${json.error_description}`);
    }

    const token = json as TokenResponse;
    this.accessToken = token.access_token;
    this.refreshToken = token.refresh_token;
    this.tokenExpiresAt = Date.now() + ((token.expires_in - 30) * 1000);
    return token;
  }

  private sendCallbackHtml(res: import('http').ServerResponse, success: boolean, message: string): void {
    const color = success ? '#22c55e' : '#ef4444';
    const icon = success ? '✓' : '✗';
    const html = `<!DOCTYPE html><html><body style="font-family:system-ui;display:flex;justify-content:center;align-items:center;height:100vh;margin:0;background:#0f172a;color:white"><div style="text-align:center"><div style="font-size:3em;color:${color}">${icon}</div><p>${message}</p></div></body></html>`;
    res.writeHead(success ? 200 : 400, { 'Content-Type': 'text/html; charset=utf-8' });
    res.end(html);
  }

  // ── Token Refresh ───────────────────────────────────────────

  /**
   * Refresh the access token using the stored refresh token.
   */
  async refreshTokenAsync(): Promise<TokenResponse> {
    if (!this.refreshToken) {
      throw new Error('No refresh token available. Please login first.');
    }

    const oidc = await this.discover();

    const body = new URLSearchParams({
      grant_type: 'refresh_token',
      client_id: this.options.oidcClientId!,
      refresh_token: this.refreshToken,
    });

    const resp = await this.fetchFn(oidc.token_endpoint!, {
      method: 'POST',
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      body: body.toString(),
    });

    const json = await resp.json();

    if (!resp.ok) {
      this.accessToken = null;
      this.refreshToken = null;
      throw new Error(`Token refresh failed: ${json.error} — ${json.error_description}`);
    }

    const token = json as TokenResponse;
    this.accessToken = token.access_token;
    this.refreshToken = token.refresh_token;
    this.tokenExpiresAt = Date.now() + ((token.expires_in - 30) * 1000);
    return token;
  }

  // ── Get Valid Token ─────────────────────────────────────────

  /**
   * Gets a valid access token, automatically refreshing if expired.
   * Throws if not logged in.
   */
  async getAccessToken(): Promise<string> {
    if (!this.accessToken) {
      throw new Error('Not logged in. Call login() first.');
    }

    if (this.isTokenExpired && this.refreshToken) {
      await this.refreshTokenAsync();
    }

    if (!this.accessToken) {
      throw new Error('Token refresh failed and no valid token available.');
    }

    return this.accessToken;
  }

  // ── Logout ──────────────────────────────────────────────────

  /**
   * Revoke the refresh token at Keycloak and clear local state.
   */
  async logout(): Promise<void> {
    if (this.refreshToken) {
      try {
        const oidc = await this.discover();
        if (oidc.revocation_endpoint) {
          const body = new URLSearchParams({
            client_id: this.options.oidcClientId!,
            token: this.refreshToken,
            token_type_hint: 'refresh_token',
          });

          await this.fetchFn(oidc.revocation_endpoint, {
            method: 'POST',
            headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
            body: body.toString(),
          });
        }
      } catch (e) {
        // Best-effort revocation
      }
    }

    this.accessToken = null;
    this.refreshToken = null;
    this.tokenExpiresAt = 0;
  }

  // ── Consent Error Handling ──────────────────────────────────

  /**
   * Parse a consent-related error from a response body or Error object.
   * Returns null if the error is not consent-related.
   */
  static parseConsentError(statusCode: number, responseBody: string): ConsentErrorInfo | null {
    if (statusCode !== 403) return null;

    try {
      const parsed = JSON.parse(responseBody);
      const errCode = parsed.error;
      
      if (
        errCode !== 'app_consent_required' &&
        errCode !== 'app_consent_pending' &&
        errCode !== 'app_consent_denied'
      ) {
        return null;
      }

      return {
        error: errCode,
        message: parsed.message,
        consentUrl: parsed.consentUrl,
        appName: parsed.appName,
        appId: parsed.appId,
      };
    } catch {
      return null;
    }
  }

  // ── State Persistence ───────────────────────────────────────

  /**
   * Restores tokens from previously saved state.
   */
  restoreTokens(accessToken: string, refreshToken: string, expiresAt: number): void {
    this.accessToken = accessToken;
    this.refreshToken = refreshToken;
    this.tokenExpiresAt = expiresAt;
  }

  /**
   * Gets the current token state for persistence.
   */
  getTokenState(): { accessToken: string | null; refreshToken: string | null; expiresAt: number } {
    return {
      accessToken: this.accessToken,
      refreshToken: this.refreshToken,
      expiresAt: this.tokenExpiresAt,
    };
  }
}
